### ISSUE TITLE
If param is not used in loss calculation, `Optimizer::update()` causes runtime panic

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#67]
Contrived example:

```rust
fn main() {
    let mut model: (Linear<5, 5>, Linear<5, 4>) = Default::default();
    let y = model.1.forward(Tensor1D::zeros().traced());
    let gradients = y.square().mean().backward();
    let mut opt: Sgd = Default::default();
    opt.update(&mut model, gradients);
}
```

**Comment by coreylowman** [comment]
Slightly worse is that you can pass a model that wasn't even used in gradients calculation into Sgd::update.

**Comment by coreylowman** [comment]
Another example with `SplitInto`:

```rust
fn main() {
    let mut model: SplitInto<(Linear<5, 4>, Linear<5, 2>)> = Default::default();
    let (_a, b) = model.forward(Tensor1D::zeros().trace());
    let loss = b.square().mean(); // NOTE: a is not used
    let gradients = loss.backwards();
    let mut opt: Sgd = Default::default();
    opt.update(&mut model, gradients);
}
```

**Comment by coreylowman** [comment]
Making owned tape generic over model pollutes *everything*. Not sure its worth it?

**Comment by coreylowman** [comment]
The root of what we really want to know for a given Gradients object is

> Does Gradients have something for each of my model parameters

The SameType check of Module::Ops and Ops gets at this in a very restrictive way.

**Comment by coreylowman** [comment]
Another fuzzy direction: Make the tape take ownership of the parameter, and then backward could reconstruct the model somehow? Something to take advantage of ownership system. Instead of trying to verify that the gradients can be used with the model, producing the gradients should require the model so they are linked somehow

**Comment by coreylowman** [comment]
FWIW pytorch silently just updates only the parameters that were involved in the computation. First need to decide what dfdx should do... fail or continue? if fail, runtime or compile time?

**Comment by vikigenius** [comment]
Interesting, although PyTorch itself silently updates only the parameters involved in the computation, I know frameworks on top of it that produce a warning about unused parameters in your model.

Returning a Result feels like a good solution since I agree this seems tricky to do in Compile time.
